### PR TITLE
Add system for checking the placement of annotations

### LINKED ISSUE #204: Annotation in the wrong location should be an error
I'm new to this library. It took me a while to debug the follow code, as it happily compiles

```
type foo = {
  foo_bar: bool <json name="fooBar">;
}
```

The annotation there has no effect. It would be nice to be warned about that.

```
$ atdgen -version
2.0.0
```

### LINKED ISSUE #227: Add a way to check if annotation is valid and/or makes sense at a given position
Inspired by https://github.com/ahrefs/atd/issues/204

For example:
```
type foo = {
  foo_bar: bool <json name="fooBar">;
}
```
It is obvious that user wanted to have the field named `fooBar` in the json output, but we have no way to report that right now. Another even more obvious case is:
```
type foo = {
  foo_bar<json nam="fooBar">: bool;
}
```
Again, `atdgen` is happy about it, while it shouldn't be ideally.

### PR DESCRIPTION
**Pull request by mjambon**
This catches misplaced annotations (typically those applying to field names and variant names) for the selected target(s). This validation is done for atdgen (ocaml, bucklescript, json, biniou) and atdpy (python/json). This caught a bug in an existing ATD test file!

Closes #204
Closes #227  

### PR checklist

- [x] New code has tests to catch future regressions
- [x] Documentation is up-to-date
- [x] `CHANGES.md` is up-to-date

### CODE REVIEW

File: atdgen/test/spec_js/spec.atd [hunk 1]
```diff
@@ -22,7 +22,7 @@ type v1 =
   [ V1
   | V2 <json name="v22">
   | V3 of string
-  | V4 of int <json name="v44">
+  | V4 <json name="v44"> of int
   | V5 of bool option]
```
Comment (mjambon) [inline atdgen/test/spec_js/spec.atd:26]: This is the bug that was caught with the validator.

**Comment by mjambon** [comment]
Thoughts anyone?

The only issue I can think of is that some existing builds will fail if some atd files contain invalid, useless annotations. The solution in that case is to remove the useless annotations. I don't think there's much value in adding a command-line option to disable validation.

**Review by cyberhuman — APPROVED** [review]

**Comment by cyberhuman** [comment]
Nice!

> The only issue I can think of is that some existing builds will fail if some atd files contain invalid, useless annotations. The solution in that case is to remove the useless annotations. I don't think there's much value in adding a command-line option to disable validation.

It sounds good. Thinking about a possible solution, the error might optionally be a warning instead. But it seems not very useful.

**Comment by cyberhuman** [comment]
@mjambon thanks for implementing this!

**Pull request merged**